### ISSUE TITLE
Regression: Huge slowdown with Trumpet winsock

### ISSUE THREAD

**Issue by stefand** [stsp/dosemu2#979]
Since commit dafd131b122541e04d80a8c9ee52efff67a815d4 dosemu2 hangs a lot when any network traffic is running through trumpet winsock. It writes the line 

ERROR: coopth: nested=2 (expected 1)

Set up dosemu networking - I am using a tap device, enable the packet driver
run winpkt
run windows
ping something with trumpet ping with a short ping interval, e.g. 1 ms

Now your windows UI is largely frozen (although it still responds every other second, so you can stop the ping) and dosemu writes the above coopth line repeatedly.

[boot.log.gz](https://github.com/stsp/dosemu2/files/4025734/boot.log.gz)

**Comment by stsp** [comment]
-D9+M please.

**Comment by stefand** [comment]
Here you are :-)
[boot.log.gz](https://github.com/stsp/dosemu2/files/4026252/boot.log.gz)

**Comment by stsp** [comment]
Could you please try this:
```
diff --git a/src/dosext/dpmi/dpmi.c b/src/dosext/dpmi/dpmi.c
index b95719204..5ca12da7c 100644
--- a/src/dosext/dpmi/dpmi.c
+++ b/src/dosext/dpmi/dpmi.c
@@ -384,6 +384,8 @@ static void dpmi_set_pm(int pm)
   dpmi_pm = pm;
   if (pm)
     run_dpmi();
+  else
+    coopth_run();      // allow dpmi thread to exit
 }
 
 int dpmi_is_valid_range(dosaddr_t addr, int len)
```

**Comment by stefand** [comment]
Works here with Netscape and ping.

**Comment by stefand** [comment]
What's the coopth stuff good for by the way? The comment in coopth.h says "cooperative multi threading between dos and dosemu". Is it for things like querying X11 events while vm86 code is running without having two Linux threads? Does it have anything to do with multiple dpmi processes / threads (if that even exists?)? Why not use proper Linux threads?

**Comment by stsp** [comment]
> Is it for things like querying X11 events while vm86 code is running without having two Linux
> threads?

No, X11 events are queried in the
real separate thread.

> Why not use proper Linux threads?

We do when that makes sense, like for querying
the X11 events, for sound and video rendering -
this is why dosemu2 is much faster than dosemu1
that was single-threaded.
But coopth serves a completely different purpose.
It allows you to "intermix" the DOS and hypervisor
code flow in any way you want. For example you
can call DOS funcs from hypervisor, or call hypervisor
funcs from DOS, and in that call-back (from DOS)
you can sleep in hypervisor and DOS interrupts
will still proceed, and only the caller DOS thread will
sleep, etc. This used to produce a weird effects,
like mouse in win31 still running if you put dosemu2
on pause with Ctrl-Alt-p or with dosdebug, and I
had to "fix" such instances by hands, because
pause means full pause, not just of 1 DOS thread.
coopth really provides you with a great power, eg
you can write any bios funcs in hypervisor space,
and they can block and sleep but the rest will work,
including DOS interrupts.
Also fdpp is running in a separate coopthread and
is completely invisible to the main dosemu2 thread.

So in short: real threads are good to off-load the
heavy computational tasks. Coopthreads are good
to integrate bios/int14/plugins into a hypervisor and
debug them with gdb in prot mode, rather than in
real mode asm.
DPMI uses coopthreads for another reason though.
In fact, there are 2 different types of coopthreads,
DPMI uses the "detached" type, as I called it. It is
not attached to any DOS context, and is therefore
just a nice way of doing continuations (a big
syntactical sugar but no real benefits).

**Comment by stsp** [comment]
And yes, in the future I am planning to extend
coopthreads to allow the attaching to the prot-mode
context. Then I will be able to compile comcom32
as a plugin, rather than as a DOS app. And that
will allow to compile the DJGPP apps with clang,
and they will still work as before. Maybe I'll even
release an SDK for that. :) So far this is just an
idea. But the existence of fdpp clearly shows that
such things are possible (provided you can spend
2 years on an experiment).

**Comment by stsp** [comment]
Besides, coopthreads in dosemu is not the
new idea. Other people tried that before, too.
For example, dosemu-1.2 had another impl
of cooperative threading (1.2 was developed
separately, so no other version of dosemu1
had it in). It was written completely differently
than mine, and was never stable, but it showed
that the idea is very interesting. dosemu2 is
about reimplementing any interesting ideas
from dosemu1, so I of course coded coopthreads
too. :) (the one from 1.2 couldn't be taken
as it had gpl2-only license)

I am not familiar with wine architecture, and
I think it doesn't have any bioses or plugins
that could get the use of such thing, but perhaps
you know otherwise. :)

**Comment by stefand** [comment]
Nah, Wine doesn't have any low level BIOS stuff etc. From a Win16 point of view it implements the DLLs, software interrupts and vxd's to talk to. You can't load any sort of device driver or do any direct talking to HW (other than maybe in/out with iopl, but no IRQs, ...). We ditched our vm86 implementation years ago and just call dosbox.

There are no plugins in the dosemu sense; the way third parties usually extend Wine is by adding / replacing Wine DLLs, see e.g. https://github.com/doitsujin/dxvk . Ideally those work in the same way on Windows and Wine.

What we do have are Windows services (they work the same way as on Windows) and an ntoskrnl.exe implementation to load drivers into. Those still can't talk to HW and run in a separate userspace process. This makes some copy protection drivers happy. Many don't work, e.g. if they try to play rootkit and intercept Windows kernel internals or try to mmap a user space process's address space into the "kernel". Other things go fairly far down the stack in a Windows-like way. The USB HID stack is implemented as ntoskrnl modules and services that work the Windows way until it hits winehid.sys that will call libSDL to talk to the Linux/Mac HID APIs or /dev nodes.

**Comment by stsp** [comment]
> but no IRQs

Have you tried uio?

**Comment by stefand** [comment]
You mean as in https://www.kernel.org/doc/html/v4.12/driver-api/uio-howto.html ?

Presumably people thought about it, but from a higher level point of view talking to the actual hardware is the smaller issue. The bigger problem is implementing all the Windows driver, PnP etc infrastructure to enumerate the existing hardware, load windows drivers and announce the driver services to applications correctly. Before we even go talk to hardware it would be nice to have stuff like this working:

*) You plug in a USB device that's a standard USB-to-serial, Linux talks to it, creates /dev/ttyUSB*, we find it, create a Win32 COM port for it, add the USB device info to the USB tree, the application enumerates USB devices with the Win32 APIs, finds the device it is looking for and now starts talking to the "COM34:" port we created.

*) You plug in a USB memory stick, it gets mounted, a windows drive letter is created and the registry HW info populated with the USB info. This halfway works. You get the drive letter in Wine (assuming the Linux kernel has the necessary drivers and filesystem to mount this thing), but it is shown as a generic hard disk and not a USB stick, and certainly not with the proper USB vendor/pci ID.

*) You plug in an old joystick, an Xbox gamepad and connect a PS4 Dualshock via Bluetooth and games can talk to them with their USB HID specific stuff instead of seeing them all as no-name DirectInput joysticks. That part is actually working these days :-) .

Once the above stuff is solved one might think about loading a Windows driver for an USB Wifi card that Linux can't talk to. But you probably don't want to use that card with Wine's non-existent network stack [it is either winsock/ws2_32 in Wine or bust], and instead have a wlan0 device on the Linux side that you can configure with wpa-supplicant and NetworkManager. Doh. Next up: Someone wants to load a Windows printer driver into his cups server and print to it from Linux LibreOffice.

All this stuff is is work for years. The way Windows works here is poorly documented. Once the infrastructure thing is solved receiving an IRQ from a PCIe card is a relatively minor problem. By that time you've probably given up on adding FUSE equivalents for all kinds of device categories and migrated ntoskrnl.exe into a Linux module anyway :-)

**Comment by stsp** [comment]
Yes, I mean LPT exactly. :)
If someone wants to use the printer driver,
there are work-arounds. But if someone
wants an LPT dongle, you need uio for
irqs.

**Comment by stefand** [comment]
A Windows app that comes with e.g. a COM or LPT copy protection dongle has a decent chance of working in Wine through whatever API has to talk to LTP and COM ports - if you have read/write perms to /dev/ttyS0 or /dev/lp0. Such an app won't be able to talk to the IO ports on Windows either.

**Comment by stsp** [comment]
Sure!
But /dev/lp0 doesn't provide you with an IRQ?

**Comment by stefand** [comment]
It doesn't, but neither does CreateFile("\\.\LPT1"). Instead, the app would probably do a WriteFile() call to send a challenge to the Dongle, then ReadFile() which will block until data arrived.

**Comment by stsp** [comment]
Do you mean there is no bit-banging API
for the use with LPT dongle? Just a basic
char IO? That would be quite strange...

**Comment by stefand** [comment]
There might be bit-banging API for parports in Windows NT, but I am not aware of one. I haven't spent much time searching though.

There's an API for serial ports in dlls/kernel32/comm.c though amd dlls/kernelbase/file.c. It seems to be implemented, calls down to ntdll via DeviceIoControl, which will eventually call tcsetattr. Afaics this API works in conjunction with Read/Write file, like tcsetattr.

**Comment by stsp** [comment]
This is interesting because people have
asked about the dongle support in dosemu2.
I thought this would require uio for irq.
If your guess is right, then the simple /dev/lp0
backend can be implemented instead.

**Comment by stefand** [comment]
More googling shows that there are tutorials that tell you program your parport via direct I/O access from userspace: http://electrosofts.com/parallel/parallelwin.html . No clue re interrupt though, but maybe the char IO API gracefully handles stuff when you use inb/outb to set some other pins. No idea if you can be notified of changes by the OS on non-data pins or have to poll them.

I guess a /dev/lp0 based implementation might be worth a try. You at least get informed when there's new data without needing access to the raw interrupt. Not sure if you can control all those aux lines, or if you have to control them. That plip socket driver probably won't run :-) .

**Comment by stefand** [comment]
Ok, this is getting waaaay off topic, but I guess the question to ask is if those dongles work on one of those new-ish USB to parallel bridges. It seems those only work for printers, although some more programmable things exist: https://archive.ph/20120630214557/http://myweb.polyu.edu.hk/~06900664r/ch341epp.htm

**Comment by stefand** [comment]
One more off topic post: CONFIG_PPDEV in the linux kernel, aka "Support for user-space parallel port device drivers" appears to expose more functionality than the lpX nodes.

**Comment by stsp** [comment]
Yes, ppdev looks like the best choice, thanks.
It has bit-banging and irqs, and is probably the
best choice for dosemu2, while /dev/lpX is a
best choice for wine.
Anyway, since I have neither dongle nor even
parport, the implementation won't materialize. :)

**Comment by stefand** [comment]
I guess if one day you have two parallel port computers you can try to run http://crynwr.com/drivers/plip.zip inside dosemu via ppdev. That eliminates the need for dongle protected software.

**Comment by stsp** [comment]
What mode does it use?
I think we only emulate SPP.

**Comment by stefand** [comment]
Not sure; the Linux counterpart plip.ko mentions two modes, depending on the cable. If I understand it right it works with a standard printer port by cross-connecting 4 data pins to 4 printer status pins. There's some advanced mode where it uses all 8 data pins.

I used it between two Linux boxes once upon a time with some 486 computers I no longer have and some random "null printer" cable I bought without having a clue about its details. The DOS driver is supposedly compatible.

**Comment by stsp** [comment]
Yes, I used plip too, and in fact fixed it
in 2.4 kernel (obviously nobody used it
but myself). If it uses SPP mode then
fine, but perhaps it would want something
with DMA.

Anyway, I just implemented the /dev/lpX
support. Was not too difficult. :)
